Dask aurora (#583)
argonne-lcf/user-guides: docs/aurora/data-science/frameworks/dask.md

        --nthreads ${NTHREADS}  --local-directory /tmp/dask-workers \
        --scheduler-file ${LOCAL_DIRECTORY}/scheduler.json >> ${LOCAL_DIRECTORY}/logs/$HOSTNAME-worker.log 2>&1
else
    dask worker \
        --nthreads ${NTHREADS} --local-directory /tmp/dask-workers \
        --scheduler-file ${LOCAL_DIRECTORY}/scheduler.json >> ${LOCAL_DIRECTORY}/logs/$HOSTNAME-worker.log 2>&1
fi
```

### Start a cluster with CPU workers

Run the following commands from a compute node on Aurora to start a Dask cluster with 104 CPU workers per node:

```bash
module load frameworks
conda activate dask
NNODES=`wc -l < $PBS_NODEFILE`
mpiexec -n $(( $NNODES * 104 )) --ppn 104 ./start_dask_aurora.sh cpu 104
```

### Start a cluster with GPU workers

Run the following commands from a compute node on Aurora to start a Dask cluster with 6 GPU workers per node:

```bash
module load frameworks
conda activate dask
NNODES=`wc -l < $PBS_NODEFILE`
mpiexec -n $(( $NNODES * 6 )) --ppn 6 ./start_dask_aurora.sh gpu 6
```

## Example

In this example, we will [estimate Pi using a Monte Carlo method](https://en.wikipedia.org/wiki/Pi#Monte_Carlo_methods). 

> Paste the following python script into a file called `pi_dask_gpu.py`. 

Reviewer: ```suggestion
Paste the following Python script into a file called `pi_dask_gpu.py`. 
```

Here is a breakdown of what the script does:
1. It connects to the Dask cluster (that you should start beforehand) and prints some information including the number of workers and available memory.
1. It divides the total number of points to sample between the workers, and each worker uses its GPU to
   - generate random points uniformly inside the unit square
   - return the number of points that are inside the unit circle
1. When the results from the workers are ready, they are aggregated to compute Pi.

> 1. A total of 5 Pi calculations are performed and timed (the very first iterations will incur in initialization and warmup costs).

Reviewer: ```suggestion
1. A total of 5x Pi calculations are performed and timed. Note, the very first iterations will incur initialization and warmup costs.
```

  Run 0           Num samples: 1.04E+10           Estimate: 3.141653798           Time taken: 1.596 s
  Run 1           Num samples: 1.04E+10           Estimate: 3.141570887           Time taken: 1.354 s
  Run 2           Num samples: 1.04E+10           Estimate: 3.141651954           Time taken: 1.451 s
  Run 3           Num samples: 1.04E+10           Estimate: 3.141636617           Time taken: 0.518 s
  Run 4           Num samples: 1.04E+10           Estimate: 3.141650108           Time taken: 0.511 s
  ```

## Connect to a Dask cluster from JupyterLab

Here are the steps to start a Dask cluster and connect to it interactively from a Jupyter notebook:

- First, request an interactive job on 1 node. Print the compute node's hostname (that you get with the command `hostname`), which will be used later.
- Then, [start a Dask cluster](#start-a-cluster-with-gpu-workers) and wait about 10 seconds for the cluster to start.
- **On your local machine**, open a ssh tunnel to the compute node (`COMPUTE_NODE` is the compute node's hostname and `YOUR_ALCF_USERNAME` is your ALCF username):
  ```bash

>   ssh -t -L 23456:localhost:23456 -L 8787:localhost:8787 YOUR_ALCF_USERNAME@bastion.alcf.anl.gov ssh -t -L 23456:localhost:23456 -L 8787:localhost:8787 login.aurora.alcf.anl.gov ssh -t -L 23456:localhost:23456 -L 8787:localhost:8787 COMPUTE_NODE

Reviewer: I get nervous when I see this multi-hop SSH tunnel line in our tutorials and docs, as there are several pitfalls as we saw during the `scikit-learn` hands-on exercise back in October: https://github.com/argonne-lcf/ALCF_Hands_on_HPC_Workshop/issues/56

Are any of those pitfalls possible here, i.e. should we suggest installing the SSH keypair in `~/.ssh/authorized_keys` for the compute node jump?

Author: I agree that the long chain of ssh commands looks ugly, but I think it should be safer and simpler than giving a more general but complex set of instructions. I will ask a few people to test this command. 
I also think it'd be a good idea to add useful ssh shortcuts and example configs somewhere in the docs that we can refer to.

Reviewer: wont it fail if any of the ports in the chain are currently in use? and the hop to the compute node will fail if the keypair is not in the `~/.ssh/authorized_keys`, no?

Author: added a link to troubleshooting page

  ```

> - **On the compute node** where you land with the above ssh command, start JupyterLab:

Reviewer: ```suggestion
- **On the compute node** where you land with the above SSH command, start JupyterLab:
```

or

```suggestion
- **On the compute node** where you land with the above `ssh` command, start JupyterLab:
```

  ```bash
  module load frameworks
  conda activate dask
  jupyter lab --no-browser --port=23456
  ```

> - Copy the line starting with `http://localhost:23456/lab?token=<TOKEN>` at the end of the jupyter command's output.

Reviewer: ```suggestion
- Copy the line starting with `http://localhost:23456/lab?token=<TOKEN>` at the end of the Jupyter command's output.
```

> - **On your local machine**, open a browser window and go to that url.

Reviewer: ```suggestion
- **On your local machine**, open a browser window and go to that URL.
```
